components/brave_wallet/browser/brave_wallet_auto_pin_service.cc

```diff
@@ -0,0 +1,256 @@
+// Copyright (c) 2022 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/brave_wallet/browser/brave_wallet_auto_pin_service.h"
+
+#include "brave/components/brave_wallet/browser/pref_names.h"
+
+namespace brave_wallet {
+
+IntentData::IntentData(const BlockchainTokenPtr& token,
+                       Operation operation,
+                       absl::optional<std::string> service)
+    : token(token.Clone()), operation(operation), service(std::move(service)) {}
+
+IntentData::~IntentData() {}
+
+BraveWalletAutoPinService::BraveWalletAutoPinService(
+    PrefService* prefs,
+    BraveWalletService* brave_wallet_service,
+    BraveWalletPinService* brave_wallet_pin_service)
+    : pref_service_(prefs),
+      brave_wallet_service_(brave_wallet_service),
+      brave_wallet_pin_service_(brave_wallet_pin_service) {
+  Restore();
+  brave_wallet_service->AddTokenObserver(
+      token_observer_.BindNewPipeAndPassRemote());
+}
+
+BraveWalletAutoPinService::~BraveWalletAutoPinService() {}
+
+void BraveWalletAutoPinService::Bind(
+    mojo::PendingReceiver<mojom::WalletAutoPinService> receiver) {
+  receivers_.Add(this, std::move(receiver));
+}
+
+mojo::PendingRemote<mojom::WalletAutoPinService>
+BraveWalletAutoPinService::MakeRemote() {
+  mojo::PendingRemote<WalletAutoPinService> remote;
+  receivers_.Add(this, remote.InitWithNewPipeAndPassReceiver());
+  return remote;
+}
+
+void BraveWalletAutoPinService::OnTokenAdded(BlockchainTokenPtr token) {
+  if (!token->is_nft) {
+    return;
+  }
+  if (!IsAutoPinEnabled()) {
+    return;
+  }
+  PostPinToken(std::move(token), base::OnceCallback<void(bool)>());
+}
+
+void BraveWalletAutoPinService::OnTokenRemoved(BlockchainTokenPtr token) {
+  if (!token->is_nft) {
+    return;
+  }
+  base::EraseIf(queue_, [&token](const std::unique_ptr<IntentData>& intent) {
+    return intent->token == token;
+  });
+  PostUnpinToken(std::move(token), base::OnceCallback<void(bool)>());
+}
+
+void BraveWalletAutoPinService::Restore() {
+  brave_wallet_service_->GetAllUserAssets(base::BindOnce(
+      &BraveWalletAutoPinService::OnTokenListResolved, base::Unretained(this)));
+}
+
+void BraveWalletAutoPinService::OnTokenListResolved(
+    std::vector<BlockchainTokenPtr> token_list) {
+  bool autopin_enabled = IsAutoPinEnabled();
+  std::set<std::string> known_tokens =
+      brave_wallet_pin_service_->GetTokens(absl::nullopt);
+  for (const auto& token : token_list) {
+    if (!token->is_nft) {
+      continue;
+    }
+    auto current_token_path =
+        BraveWalletPinService::GetPath(absl::nullopt, token);
+    if (!current_token_path) {
+      continue;
+    }
+
+    auto it = known_tokens.find(current_token_path.value());
+    if (it != known_tokens.end()) {
+      known_tokens.erase(it);
+    }
+
```

Comment:
`known_tokens.erase(current_token_path.value())` is enough

Reply:
It is failing if there is no such value